r/src/array_from_vector.cpp

```diff
@@ -351,15 +351,20 @@ std::shared_ptr<Array> MakeFactorArrayImpl(Rcpp::IntegerVector_ factor,
 
 std::shared_ptr<Array> MakeFactorArray(Rcpp::IntegerVector_ factor,
                                        const std::shared_ptr<arrow::DataType>& type) {
-  SEXP levels = factor.attr("levels");
-  int n = Rf_length(levels);
-  if (n < 128) {
+  auto* dict_type = checked_cast<arrow::DictionaryType*>(type.get());
+  auto index_type = dict_type->index_type();
+  if (index_type->Equals(int8())) {
```

Comment:
Faster if written as `index_type->id() == Type::INT8`. You may or may not bother, depending on how important this function is.